blockchain/blockchain.go

```diff
@@ -564,94 +550,89 @@ func (b *Blockchain) WriteBlocks(blocks []*types.Block) error {
+		return fmt.Errorf(
+			"uncle root hash mismatch: have %s, want %s",
+			hash,
+			block.Header.Sha3Uncles,
+		)
+	}
 
-			return fmt.Errorf(
-				"transaction root hash mismatch: have %s, want %s",
-				hash,
-				block.Header.TxRoot,
-			)
-		}
+	// TODO, the wrapper around transactions
+	if hash := buildroot.CalculateTransactionsRoot(block.Transactions); hash != block.Header.TxRoot {
 
-		parent = block.Header
+		return fmt.Errorf(
+			"transaction root hash mismatch: have %s, want %s",
+			hash,
+			block.Header.TxRoot,
+		)
 	}
 
+	parent = block.Header
+
 	// Checks are passed, write the chain
-	for indx, block := range blocks {
-		header := block.Header
+	header := block.Header
 
-		// Process and validate the block
-		res, err := b.processBlock(blocks[indx])
-		if err != nil {
-			return err
-		}
+	// Process and validate the block
+	res, err := b.processBlock(block)
+	if err != nil {
+		return err
+	}
 
-		if err := b.writeBody(block); err != nil {
-			return err
-		}
+	if err := b.writeBody(block); err != nil {
+		return err
+	}
 
-		// Write the header to the chain
-		evnt := &Event{}
-		if err := b.writeHeaderImpl(evnt, header); err != nil {
-			return err
-		}
+	// Write the header to the chain
+	evnt := &Event{}
+	if err := b.writeHeaderImpl(evnt, header); err != nil {
+		return err
+	}
 
-		// write the receipts, do it only after the header has been written.
-		// Otherwise, a client might ask for a header once the receipt is valid
-		// but before it is written into the storage
-		if err := b.db.WriteReceipts(block.Hash(), res.Receipts); err != nil {
-			return err
-		}
+	// write the receipts, do it only after the header has been written.
+	// Otherwise, a client might ask for a header once the receipt is valid
+	// but before it is written into the storage
+	if err := b.db.WriteReceipts(block.Hash(), res.Receipts); err != nil {
+		return err
+	}
 
-		b.dispatchEvent(evnt)
+	b.dispatchEvent(evnt)
 
-		// Update the average gas price
-		b.UpdateGasPriceAvg(new(big.Int).SetUint64(header.GasUsed))
+	// Update the average gas price
+	b.UpdateGasPriceAvg(new(big.Int).SetUint64(header.GasUsed))
 
-		logArgs := []interface{}{
-			"number", header.Number,
-			"hash", header.Hash,
-			"txns", len(block.Transactions),
-		}
-		if prevHeader, ok := b.GetHeaderByNumber(header.Number - 1); ok {
-			diff := header.Timestamp - prevHeader.Timestamp
-			logArgs = append(logArgs, "generation_time_in_sec", diff)
-		}
-		b.logger.Info("new block", logArgs...)
+	logArgs := []interface{}{
+		"number", header.Number,
+		"hash", header.Hash,
+		"txns", len(block.Transactions),
+	}
+	if prevHeader, ok := b.GetHeaderByNumber(header.Number - 1); ok {
+		diff := header.Timestamp - prevHeader.Timestamp
+		logArgs = append(logArgs, "generation_time_in_sec", diff)
 	}
 
+	b.logger.Info("new block", logArgs...)
 	b.logger.Info("new head", "hash", b.Header().Hash, "number", b.Header().Number)
```

Comment:
i don't think we need this `new head` message any more, since the previous `new block` will log all the relevant information.

Reply:
Agreed 💯 

Fixed 👍 

a90c517